server/neptune/workflows/internal/pr/runner.go

```diff
@@ -0,0 +1,217 @@
+	ShutdownReceiver      *receiver.ShutdownReceiver
+	Scope                 workflowMetrics.Scope
+	TFWorkflow            TFWorkflow
+	TFStateReceiver       StateReceiver
+
+	// mutable state
+	state                 RunnerState
+	lastAttemptedRevision string
+}
+
+func newRunner(ctx workflow.Context, scope workflowMetrics.Scope, tfWorkflow TFWorkflow, internalNotifiers []WorkflowNotifier) *Runner {
+	revisionReceiver := receiver.NewRevisionReceiver(ctx, scope)
+	shutdownReceiver := receiver.NewShutdownReceiver(ctx, scope)
+	return &Runner{
+		RevisionSignalChannel: workflow.GetSignalChannel(ctx, receiver.TerraformRevisionSignalID),
+		RevisionReceiver:      &revisionReceiver,
+		ShutdownSignalChannel: workflow.GetSignalChannel(ctx, receiver.ShutdownSignalID),
+		ShutdownReceiver:      &shutdownReceiver,
+		Scope:                 scope,
+		TFWorkflow:            tfWorkflow,
+		TFStateReceiver:       StateReceiver{InternalNotifiers: internalNotifiers},
+	}
+}
+
+// Run handles managing the workflow's context lifecycles as new signals/poll events are received and
+// change the current PRAction status
+func (r *Runner) Run(ctx workflow.Context) error {
+	var action Action
+	var prRevision receiver.Revision
+
+	//TODO: add approve signal, timeouts, poll variation of shutdown signal
+	s := workflow.NewSelector(ctx)
+	s.AddReceive(r.RevisionSignalChannel, func(c workflow.ReceiveChannel, more bool) {
+		prRevision = r.RevisionReceiver.Receive(c, more)
+		action = onNewRevision
+	})
+
+	s.AddReceive(r.ShutdownSignalChannel, func(c workflow.ReceiveChannel, more bool) {
+		r.ShutdownReceiver.Receive(c, more)
+		action = onShutdown
+		r.state = complete
+	})
+
+	_, cancel := workflow.WithCancel(ctx)
+	for {
+		s.Select(ctx)
+		switch action {
+		case onNewRevision:
+			revisionCtx := workflow.WithValue(ctx, internalContext.SHAKey, prRevision.Revision)
+			workflow.GetLogger(revisionCtx).Info("received revision")
+			if process := r.shouldProcessRevision(prRevision); !process {
+				continue
+			}
+			// cancel in progress workflow (if it exists) and start up a new one
+			cancel()
+			revisionCtx, cancel = workflow.WithCancel(revisionCtx)
+			r.state = working
+			r.lastAttemptedRevision = prRevision.Revision
+			workflow.Go(revisionCtx, func(c workflow.Context) {
+				r.processRevision(c, prRevision)
+			})
+		case onShutdown:
+			//todo: maybe optimize by cancelling in progress child workflows
+			workflow.GetLogger(ctx).Info("received shutdown request")
+			return nil
+		}
+	}
+}
+
+func (r *Runner) shouldProcessRevision(prRevision receiver.Revision) bool {
+	// ignore reruns when revision is still in progress
+	if r.lastAttemptedRevision == prRevision.Revision && r.state != waiting {
+		return false
+	}
+	return true
+}
+
+// processRevision handles spinning off child Terraform workflows per root and
+// dealing with any failed policies by reviewing set of approvals
+func (r *Runner) processRevision(ctx workflow.Context, prRevision receiver.Revision) {
```

Comment:
Theres a lot going on in one struct, consider just having the runner manage the entry point and have something like a RevisionHandler/RevisionProcessor handle the rest of the logic. Testing is going to be a nightmare otherwise in addition to making future changes.

Reply:
I set up a RevisionProcessor to break up the changes